bittide/src/Bittide/Counter.hs

```diff
@@ -0,0 +1,134 @@
+  deriving (Generic, NFDataX)
+
+-- | Indicates whether 'domainDiffCounter' is actively counting or still in reset.
+type Active = Bool
+
+-- | Determine speed differences between two domains. If the source domain is
+-- faster than the destination domain, the result will become smaller. Vice versa,
+-- if the source domain is slower than the destination domain, the result will
+-- become larger. This is analogous to what would happen to a FIFO's data count
+-- when continuously written to by the source domain and read from by the
+-- destination domain. To ease integration in control algorithms, this component
+-- makes sure it starts counting at zero. It also waits for the incoming counter
+-- to become active, i.e. non-zero, before starting to count itself.
+--
+-- If both domains support initial values, 'domainDiffCounter' does not need to
+-- be reset.
+--
+-- To reset this component, the reset should be asserted for at least one cycle in
+-- the source domain _plus_ four cycles in the destination domain. The reset in the
+-- destination domain should be deasserted at the same time or *after* the one in
+-- the source domain for glitchless operation.
+--
+-- __N.B.__:
+--   This function will only work properly if the given domains are pretty close
+--   to another for a number of reasons:
+--
+--     1. It uses an 8-bit Gray counter internally for CDC
+--     2. It uses one 64-bit counter in each domain
+--     2. Its output is constrained to @Signed 16@
+--
+domainDiffCounter ::
+  forall src dst .
+  ( KnownDomain src
+  , KnownDomain dst
+  ) =>
+  Clock src -> Reset src ->
+  Clock dst -> Reset dst ->
+  -- | Counter and boolean indicating whether the component is currently active
+  Signal dst (Signed 16, Active)
+domainDiffCounter clkSrc rstSrc clkDst rstDst =
+  mealy clkDst rstDst enableGen go DdcInReset counter
+ where
+  -- 64 bits is enough for approximately 3 millenia @ 200 MHz
+  counter = synchronizedSuccCounter @64 clkSrc rstSrc clkDst rstDst
+
+  go :: DdcState -> Unsigned 64 -> (DdcState, (Signed 16, Bool))
+  go DdcInReset c1
+    | c1 == 0   = (DdcInReset,          (0, False))
+    | otherwise = (DdcRunning (c1 + 1), (0, True))
+  go (DdcRunning c0) c1 = (DdcRunning (c0 + 1), (c0 `subAndTruncate` c1, True))
+
+  subAndTruncate :: Unsigned 64 -> Unsigned 64 -> Signed 16
+  subAndTruncate c0 c1 = truncateB (unsignedToSigned c0 - unsignedToSigned c1)
+
+-- | A counter that counts /up/, synchronized from the domain @src@ to domain @dst@. To
+-- reset this component, the reset should be asserted for at least one cycle in the
+-- source domain _plus_ four cycles in the destination domain.
+--
+-- __N.B.__: This function uses an 8-bit Gray counter internally, and will therefore
+--           only work properly if both clock speeds are pretty close to one another.
+synchronizedSuccCounter ::
+  forall n m src dst .
+  ( KnownDomain src
+  , KnownDomain dst
+  , KnownNat n
+  , 8 <= n
+
+  -- TODO: Haxiom this out:
```

Comment:
Don't forget.

Reply:
Apparently this was only necessary in earlier iterations of the code :shrug: